File: pysrc/fractal/mind/bicameral/_op/evaluate.py
"""
Evaluate
========

Evaluate a single neural network for a one-player game.

"""

from fractal.world import Nature
from fractal.brain import Brain
from fractal import Fractal

from .._mind import Mind


class Evaluate(
    Mind,
):
    @classmethod
    def thinking(
        cls,
        nature: Nature,
        brain: Brain,
        evaluation: Evaluation,
        handler: AtDusk,
    ):
        make_oracles = brain.copy(
            on_gpu=Fractal.Mind.Imagination.use_gpu,
            test_mode=true,
        )
        # simulator = Simulator(make_oracles, record_trace) do oracle
        # simulator.MctsPlayer(gspec, oracle, params.mcts)

        # samples = simulate()
        # simulator, gspec, params.sim,
        # game_simulated=(() -> Handlers.checkpoint_game_played(handler))
        # return rewards_and_redundancy(samples, gamma=params.mcts.gamma,)

File: src/fractal/mind/_op/evaluate.rs

pub trait Evaluate {
    fn evaluate(&self) {}
    //! Evaluate a single brain for a unicameral mind.
}

impl Evaluate for Mind (
    fn evaluate(&self) {
        brain = &self.brain.clone(
            on_gpu=Fractal.Mind.Imagination.use_gpu,
            test_mode=true,
        );

        imagination = Imagination(make_oracles, record_trace);
        imagination.summon(nature: Nature, demon: Demon, foresight: Foresight);

        visions = &self.imagination.imagine()
        # simulator, gspec, params.sim,
        imagined  =(() -> Checkpoints.imagined_the_future(&self.awareness))
        return (imagined, skepticism=&self.imagination.skepticism,)
        }
    }
}
